### PR TITLE
Adding integration test for ClickHouse.Client

### PR DESCRIPTION
**Pull request by DarkWanderer**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Build/Testing Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added integration test for external .NET client ([ClickHouse.Client](https://github.com/DarkWanderer/ClickHouse.Client))

Additional info:
- I don't have a machine to run the integration test suite locally, so this PR is a bit of 'shot in the dark'

### CODE REVIEW

File: docker/test/integration/dotnet_client/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,8 @@
+FROM mcr.microsoft.com/dotnet/sdk:3.1
```
Comment (qoega) [inline docker/test/integration/dotnet_client/Dockerfile:1]: Add a comment how to build an image + it helps to find its name.

File: tests/integration/test_mysql_protocol/test.py [hunk 2]
```diff
@@ -416,6 +422,15 @@ def test_golang_client(server_address, golang_container):
     assert stdout == reference
 
 
+def test_dotnet_client(server_address, dotnet_container):
+    code, (stdout, stderr) = dotnet_container.exec_run(
+        'dotnet run --host {host} --port {port} --user default --password 123 --database default'
+        .format(host=server_address, port=server_port), demux=True)
```
Comment (qoega) [inline tests/integration/test_mysql_protocol/test.py:428]: Why do you add it to mysql_protocol test? Which protocol is used? server_port is a native one.

Comment (DarkWanderer) [reply tests/integration/test_mysql_protocol/test.py:428]: I've used golang client as base example, it's in same file. What is the variable with http port?

**Comment by alexey-milovidov** [comment]
@DarkWanderer This test still requires attention.

**Comment by DarkWanderer** [comment]
I do not understand how to fix this without further input

**Comment by alexey-milovidov** [comment]
We need this test :)

**Comment by alexey-milovidov** [comment]
You can check the logs for `Integration tests flaky check (asan)`

**Comment by alexey-milovidov** [comment]
@mergifyio update

**Comment by alexey-milovidov** [comment]
```
>               raise Exception(f"Command {args} return non-zero code {res.returncode}: {res.stderr.decode('utf-8')}")
E               Exception: Command ['docker-compose', '-p', 'roottestmysqlprotocol', '-f', '/compose/docker_compose_dotnet_client.yml', 'up', '--no-recreate', '-d', '--no-build'] return non-zero code 1: Building with native build. Learn about native build in Compose here: https://docs.docker.com/go/compose-native-build/
E               Found orphan containers (roottestmysqlprotocol_mysql_client_1, roottestmysqlprotocol_node_1, roottestmysqlprotocol_golang1_1) for this project. If you removed or renamed this service in your compose file, you can run this command with the --remove-orphans flag to clean it up.
E               Pulling dotnet1 (yandex/clickhouse-dotnet-client:latest)...
E               The image for the service you're trying to recreate has been removed. If you continue, volume data could be lost. Consider backing up your data before continuing.
E               
E               pull access denied for yandex/clickhouse-dotnet-client, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
```

**Comment by alexey-milovidov** [comment]
```
$ docker pull yandex/clickhouse-dotnet-client
Using default tag: latest
Error response from daemon: pull access denied for yandex/clickhouse-dotnet-client, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
```

What is the real location of this Docker repository?

**Comment by alexey-milovidov** [comment]
There are no similar images: https://hub.docker.com/search?q=clickhouse-dotnet-client&type=image

**Comment by DarkWanderer** [comment]
Hi @alexey-milovidov,

When I asked about this in Telegram dev chat, I've been told the following:

> English: Good afternoon. I saw the above mention of integration tests for third-party clients. Can I add my own library? I understand correctly what is needed for this:
> * add an image with tests in docker/test/integration 
> * add .yml in docker/test/integration/runner/compose
> ?

> English: Yes, that's right. It is necessary that a fixed version of your library is taken there and we will test compatibility with it (we will then change the version manually when necessary).

The only way I see it working is if the image is built and pushed by Yandex CI itself, into Yandex-owned namespace. If the container is in namespace which I own, obviously you cannot change version and push new image

Of course I can try and create a docker feed which will post exclusively tagged versions of test images so you tefer to a specific image. But that seems quite kludgy.

What would you suggest?

**Comment by alexey-milovidov** [comment]
Maybe better to integrate build of .NET client inside integration tests Docker image (instead of creating another image).

**Comment by DarkWanderer** [comment]
Okay, it might work
- which files should I look at? Where is this image defined?
- what are the environment variables I can refer to for connextion details? (host/port/username/password)

**Comment by alexey-milovidov** [comment]
1. `docker/test/integration`
2. Do it similarly to other integration tests.

File: tests/integration/test_mysql_protocol/test.py (hunk 2 above)
Comment (alexey-milovidov) [reply tests/integration/test_mysql_protocol/test.py]: `test_mysql_protocol` has `test_golang_client` to test MySQL compatibility protocol of ClickHouse by using golang MySQL driver.

But your .NET driver does not use MySQL compatibility protocol. So, better to make a separate test directory for it.

File: docker/test/integration/dotnet_client/Program.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        }
+
+        private static string selectSql = @"SELECT NULL, toInt8(-8), toUInt8(8), toInt16(-16), toUInt16(16), toInt16(-32), toUInt16(32), toInt64(-64), toUInt64(64), toFloat32(32e6), toFloat32(-32e6), toFloat64(64e6), toFloat64(-64e6), 'TestString', toFixedString('ASD',3), toFixedString('ASD',5), toUUID('00000000-0000-0000-0000-000000000000'), toUUID('61f0c404-5cb3-11e7-907b-a6006ad3dba0'), toIPv4('1.2.3.4'), toIPv4('255.255.255.255'), CAST('a', 'Enum(\'a\' = 1, \'b\' = 2)'), CAST('a', 'Enum8(\'a\' = -1, \'b\' = 127)'), CAST('a', 'Enum16(\'a\' = -32768, \'b\' = 32767)'), array(1, 2, 3), array('a', 'b', 'c'), array(1, 2, NULL), toInt32OrNull('123'), toInt32OrNull(NULL), CAST(NULL AS Nullable(DateTime)), CAST(NULL AS LowCardinality(Nullable(String))), toLowCardinality('lowcardinality'), tuple(1, 'a', 8), tuple(123, tuple(5, 'a', 7)), toDateOrNull('1999-11-12'), toDateTime('1988-08-28 11:22:33'), toDateTime64('2043-03-01 18:34:04.4444444', 9), toDecimal32(123.45, 3), toDecimal32(-123.45, 3), toDecimal64(1.2345, 7), toDecimal64(-1.2345, 7), toDecimal128(12.34, 9), toDecimal128(-12.34, 9), toIPv6('2001:0db8:85a3:0000:0000:8a2e:0370:7334')";
```
Comment (alexey-milovidov) [inline docker/test/integration/dotnet_client/Program.cs:71]: > Program.cs(71,31): warning CS0414: The field 'Program.selectSql' is assigned but its value is never used [/client/clickhouse.test.csproj]

**Comment by alexey-milovidov** [comment]
Continued in #33368.

**Comment by alexey-milovidov** [comment]
@DarkWanderer Congratulations, the test is merged!

**Comment by DarkWanderer** [comment]
Thanks for picking this up @alexey-milovidov :+1:

**Pull request closed without merging**